packages/devtools/lib/src/timeline/timeline_protocol.dart

```diff
@@ -166,120 +169,201 @@ class TimelineData {
-      );
-      e.parent = parentEvent;
-      _asyncEvents[asyncUID] = e;
+  /// Looks through [_pendingEvents] and attempts to add events to frames in
+  /// [_pendingFrames].
+  void _maybeAddPendingEvents() {
+    final List<TimelineFrame> frames = _getAndSortWellFormedFrames();
+    for (TimelineFrame frame in frames) {
+      // Sort _pendingEvents by their startTime. This ensures we will add the
+      // first matching event within the time boundary to the frame.
+      _pendingEvents.sort((a, b) => a.startTime.compareTo(b.startTime));
+
+      // Make a copy of `_pendingEvents` to iterate through.
+      final List<TimelineEvent> events =
+          List<TimelineEvent>.from(_pendingEvents);
+
+      for (TimelineEvent event in events) {
+        final bool eventAdded = _maybeAddEventToFrame(event, frame);
+        if (eventAdded) _pendingEvents.remove(event);
+      }
     }
   }
 
-  void _handleAsyncInstantEvent(TraceEvent event) {
-    final String asyncUID = event.asyncUID;
+  /// Add event to an available frame in [_pendingFrames] if we can, or otherwise add it
+  /// to [_pendingEvents].
+  void _maybeAddEvent(TimelineEvent event) {
+    if (!event.isPipelineProduceFlow && !event.isPipelineConsumeFlow) {
+      // We do not care about other events.
+      return;
+    }
 
-    final TimelineEvent e = TimelineEvent(
-      event.name,
-      event.timestampMicros,
-      _getTimelineEventType(event),
-    );
-    e.endTime = event.timestampMicros + event.duration;
+    bool eventAdded = false;
 
-    final TimelineEvent parent = _asyncEvents[asyncUID];
-    if (parent != null) {
-      e.parent = parent;
+    final List<TimelineFrame> frames = _getAndSortWellFormedFrames();
+    for (TimelineFrame frame in frames) {
+      eventAdded = _maybeAddEventToFrame(event, frame);
     }
+
+    if (!eventAdded) _pendingEvents.add(event);
   }
 
-  void _handleAsyncEndEvent(TraceEvent event) {
-    final String asyncUID = event.asyncUID;
+  /// Add event `e` to frame `f` if it meets the necessary criteria.
+  ///
+  /// Returns a bool indicating whether the event was added to the frame.
+  bool _maybeAddEventToFrame(TimelineEvent e, TimelineFrame f) {
+    assert(f.wellFormed);
 
-    final TimelineEvent current = _asyncEvents[asyncUID];
-    if (current != null) {
-      current.endTime = event.timestampMicros;
-      _asyncEvents[asyncUID] = current.parent;
+    // TODO(kenzie): consider trimming VSYNC layer from pipelineProduceFlow. It
+    // can start outside of the frame's time boundaries and could pose a risk
+    // for us missing a frame.
 
-      if (_asyncEvents[asyncUID] == null) {
-        frames[event.frameId].addEvent(current);
-      }
+    // Ensure the event fits within the frame's time boundaries.
+    if (!_eventOccursWithinFrameBoundaries(e, f)) return false;
+
+    bool eventAdded = false;
+
+    if (e.isPipelineProduceFlow && f.pipelineProduceFlow == null) {
+      f.pipelineProduceFlow = e;
```

Comment:
the bodies of these two if statements are the same. merge the if statements.

Reply:
The bodies are actually subtlety different. One is 
`f.pipelineProduceFlow = e;` and the other is
`f.pipelineConsumeFlow = e;`.

Reply:
maybe we should make the variable names shorter
if these where
f.produceFlow
and
f.consumeFlow 
I probably wouldn't have made the mistake :)

Reply:
better yet, renamed them to cpuEventFlow and gpuEventFlow